File: linux/src/protocol/ESP32Protocol.cpp
// FAME Smart Flasher - Linux Qt Port
// Copyright 2025 Fyrby Additive Manufacturing & Engineering
// SPDX-License-Identifier: Proprietary

#include "ESP32Protocol.h"

namespace {

// Helper to append little-endian 16-bit value
void appendLE16(QByteArray& data, uint16_t value)
{
    data.append(static_cast<char>(value & 0xFF));
    data.append(static_cast<char>((value >> 8) & 0xFF));
}

// Helper to append little-endian 32-bit value
void appendLE32(QByteArray& data, uint32_t value)
{
    data.append(static_cast<char>(value & 0xFF));
    data.append(static_cast<char>((value >> 8) & 0xFF));
    data.append(static_cast<char>((value >> 16) & 0xFF));
    data.append(static_cast<char>((value >> 24) & 0xFF));
}

// Helper to read little-endian 16-bit value
uint16_t readLE16(const QByteArray& data, int offset)
{
    return static_cast<uint16_t>(static_cast<uint8_t>(data[offset])) |
           (static_cast<uint16_t>(static_cast<uint8_t>(data[offset + 1])) << 8);
}

// Helper to read little-endian 32-bit value
uint32_t readLE32(const QByteArray& data, int offset)
{
    return static_cast<uint32_t>(static_cast<uint8_t>(data[offset])) |
           (static_cast<uint32_t>(static_cast<uint8_t>(data[offset + 1])) << 8) |
           (static_cast<uint32_t>(static_cast<uint8_t>(data[offset + 2])) << 16) |
           (static_cast<uint32_t>(static_cast<uint8_t>(data[offset + 3])) << 24);
}

/**
 * Build a command packet (before SLIP encoding)
 * @param command Command opcode
 * @param payload Command payload
 * @param checksum Checksum value (for data commands)
 * @return Raw command packet
 */
QByteArray buildPacket(ESP32Command command, const QByteArray& payload, uint32_t checksum = 0)
{
    QByteArray packet;
    packet.reserve(8 + payload.size());

    // Direction: 0x00 for request
    packet.append(static_cast<char>(0x00));
    // Command opcode
    packet.append(static_cast<char>(static_cast<uint8_t>(command)));
    // Data size (little-endian 16-bit)
    appendLE16(packet, static_cast<uint16_t>(payload.size()));
    // Checksum (little-endian 32-bit)
    appendLE32(packet, checksum);
    // Payload
    packet.append(payload);

    return packet;
}

} // anonymous namespace

std::optional<ESP32Response> ESP32Response::parse(const QByteArray& packet)
{
    if (packet.size() < 8) {
        return std::nullopt;
    }

    uint8_t direction = static_cast<uint8_t>(packet[0]);
    // Response direction should be 0x01
    if (direction != 0x01) {
        return std::nullopt;
    }

    ESP32Response response;
    response.direction = direction;
    response.command = static_cast<uint8_t>(packet[1]);
    response.size = readLE16(packet, 2);
    response.value = readLE32(packet, 4);

    int dataEndIndex = qMin(8 + static_cast<int>(response.size), packet.size());
    if (packet.size() > 8) {
        response.data = packet.mid(8, dataEndIndex - 8);
    }

    // Status bytes are at the START of the data section (not the end!)
    // Format: [status (1 byte)][error (1 byte)][optional additional data]
    response.status = response.data.size() >= 1 ? static_cast<uint8_t>(response.data[0]) : 0;
    response.error = response.data.size() >= 2 ? static_cast<uint8_t>(response.data[1]) : 0;

    return response;
}

namespace ESP32Protocol {

uint32_t calculateChecksum(const QByteArray& data)
{
    uint8_t checksum = CHECKSUM_SEED;
    for (int i = 0; i < data.size(); ++i) {
        checksum ^= static_cast<uint8_t>(data[i]);
    }
    return static_cast<uint32_t>(checksum);
}

QByteArray buildSyncCommand()
{
    QByteArray payload;
    payload.append(static_cast<char>(0x07));
    payload.append(static_cast<char>(0x07));
    payload.append(static_cast<char>(0x12));
    payload.append(static_cast<char>(0x20));

    // 32 bytes of 0x55
    for (int i = 0; i < 32; ++i) {
        payload.append(static_cast<char>(0x55));
    }

    return buildPacket(ESP32Command::Sync, payload);
}

QByteArray buildSpiAttachCommand(uint32_t config)
{
    QByteArray payload;
    // SPI configuration - 0 means use default SPI flash pins
    appendLE32(payload, config);
    // For ESP32-C3, we need 8 bytes total (second word is also 0)
    appendLE32(payload, 0);
    return buildPacket(ESP32Command::SpiAttach, payload);
}

QByteArray buildFlashBeginCommand(
    uint32_t size,
    uint32_t numBlocks,
    uint32_t blockSize,
    uint32_t offset,
    bool encrypted)
{
    QByteArray payload;
    payload.reserve(20);  // 5 x 32-bit words for ROM loader

    // Erase size (little-endian)
    appendLE32(payload, size);
    // Number of blocks
    appendLE32(payload, numBlocks);
    // Block size
    appendLE32(payload, blockSize);
    // Offset
    appendLE32(payload, offset);
    // Encryption flag (ROM loader requires this 5th word)
    // 0 = not encrypted, 1 = encrypted
    appendLE32(payload, encrypted ? 1 : 0);

    return buildPacket(ESP32Command::FlashBegin, payload);
}

QByteArray buildFlashDataCommand(const QByteArray& blockData, uint32_t sequenceNumber)
{
    QByteArray payload;
    payload.reserve(16 + blockData.size());

    // Data length (little-endian)
    appendLE32(payload, static_cast<uint32_t>(blockData.size()));
    // Sequence number
    appendLE32(payload, sequenceNumber);
    // Reserved (8 bytes of zeros)
    for (int i = 0; i < 8; ++i) {
        payload.append(static_cast<char>(0));
    }
    // Actual data
    payload.append(blockData);

    uint32_t checksum = calculateChecksum(blockData);
    return buildPacket(ESP32Command::FlashData, payload, checksum);
}

QByteArray buildFlashEndCommand(bool reboot)
{
    QByteArray payload;
    // 0 = reboot, 1 = stay in bootloader
    uint32_t flag = reboot ? 0 : 1;
    appendLE32(payload, flag);

    return buildPacket(ESP32Command::FlashEnd, payload);
}

QByteArray buildChangeBaudCommand(uint32_t newBaud, uint32_t oldBaud)
{
    QByteArray payload;
    appendLE32(payload, newBaud);
    appendLE32(payload, oldBaud);

    return buildPacket(ESP32Command::ChangeBaudRate, payload);
}

QByteArray buildReadRegCommand(uint32_t address)
{
    QByteArray payload;
    appendLE32(payload, address);
    return buildPacket(ESP32Command::ReadReg, payload);
}

QByteArray buildWriteRegCommand(uint32_t address, uint32_t value, uint32_t mask, uint32_t delayUs)
{
    QByteArray payload;
    appendLE32(payload, address);
    appendLE32(payload, value);
    appendLE32(payload, mask);
    appendLE32(payload, delayUs);
    return buildPacket(ESP32Command::WriteReg, payload);
}

} // namespace ESP32Protocol

File: windows/Protocol/ESP32Protocol.cs
// FAME Smart Flasher - C# Port
// Copyright 2025 Fyrby Additive Manufacturing & Engineering
// SPDX-License-Identifier: Proprietary

using System.Buffers.Binary;

namespace FAMESmartFlasher.Protocol;

/// <summary>
/// ESP32 bootloader command opcodes
/// </summary>
public enum ESP32Command : byte
{
    Sync = 0x08,
    FlashBegin = 0x02,
    FlashData = 0x03,
    FlashEnd = 0x04,
    ChangeBaudRate = 0x0F,
    ReadReg = 0x0A,
    WriteReg = 0x09,
    SpiAttach = 0x0D
}

/// <summary>
/// ESP32-C3 register addresses for watchdog control
/// </summary>
public static class ESP32C3Registers
{
    public const uint RtcCntlBase = 0x60008000;

    // RTC Watchdog Config
    public const uint RtcWdtConfig0 = RtcCntlBase + 0x0090;
    public const uint RtcWdtWprotect = RtcCntlBase + 0x00A8;
    public const uint RtcWdtWkey = 0x50D83AA1;

    // Super Watchdog Config
    public const uint SwdConf = RtcCntlBase + 0x00AC;
    public const uint SwdWprotect = RtcCntlBase + 0x00B0;
    public const uint SwdWkey = 0x8F1D312A;

    // Bit positions
    public const uint WdtEnBit = 1u << 31;
    public const uint SwdAutoFeedEnBit = 1u << 31;
    public const uint SwdDisableBit = 1u << 30;
}

/// <summary>
/// ESP32 protocol packet builder
/// </summary>
public static class ESP32Protocol
{
    /// <summary>
    /// Checksum seed value
    /// </summary>
    public const byte ChecksumSeed = 0xEF;

    /// <summary>
    /// Default block size for flash data
    /// </summary>
    public const int FlashBlockSize = 1024;

    /// <summary>
    /// Calculate XOR checksum for data
    /// </summary>
    /// <param name="data">Data to checksum</param>
    /// <returns>Checksum value</returns>
    public static uint CalculateChecksum(byte[] data)
    {
        byte checksum = ChecksumSeed;
        foreach (var b in data)
        {
            checksum ^= b;
        }
        return checksum;
    }

    /// <summary>
    /// Build a command packet (before SLIP encoding)
    /// </summary>
    private static byte[] BuildPacket(ESP32Command command, byte[] data, uint checksum = 0)
    {
        var packet = new List<byte>(8 + data.Length);

        // Direction: 0x00 for request
        packet.Add(0x00);
        // Command opcode
        packet.Add((byte)command);
        // Data size (little-endian 16-bit)
        packet.Add((byte)(data.Length & 0xFF));
        packet.Add((byte)((data.Length >> 8) & 0xFF));
        // Checksum (little-endian 32-bit)
        packet.Add((byte)(checksum & 0xFF));
        packet.Add((byte)((checksum >> 8) & 0xFF));
        packet.Add((byte)((checksum >> 16) & 0xFF));
        packet.Add((byte)((checksum >> 24) & 0xFF));
        // Payload
        packet.AddRange(data);

        return packet.ToArray();
    }

    /// <summary>
    /// Build SYNC command packet
    /// SYNC payload: 0x07 0x07 0x12 0x20 followed by 32 bytes of 0x55
    /// </summary>
    public static byte[] BuildSyncCommand()
    {
        var payload = new List<byte> { 0x07, 0x07, 0x12, 0x20 };
        for (int i = 0; i < 32; i++)
        {
            payload.Add(0x55);
        }
        return BuildPacket(ESP32Command.Sync, payload.ToArray());
    }

    /// <summary>
    /// Build SPI_ATTACH command packet
    /// Required before FLASH_BEGIN when using ROM bootloader (not stub)
    /// </summary>
    /// <param name="config">SPI configuration (0 = use default pins)</param>
    public static byte[] BuildSpiAttachCommand(uint config = 0)
    {
        var payload = new byte[8];
        BinaryPrimitives.WriteUInt32LittleEndian(payload.AsSpan(0, 4), config);
        BinaryPrimitives.WriteUInt32LittleEndian(payload.AsSpan(4, 4), 0);
        return BuildPacket(ESP32Command.SpiAttach, payload);
    }

    /// <summary>
    /// Build FLASH_BEGIN command packet
    /// </summary>
    public static byte[] BuildFlashBeginCommand(
        uint size,
        uint numBlocks,
        uint blockSize,
        uint offset,
        bool encrypted = false)
    {
        var payload = new byte[20]; // 5 x 32-bit words for ROM loader

        BinaryPrimitives.WriteUInt32LittleEndian(payload.AsSpan(0, 4), size);
        BinaryPrimitives.WriteUInt32LittleEndian(payload.AsSpan(4, 4), numBlocks);
        BinaryPrimitives.WriteUInt32LittleEndian(payload.AsSpan(8, 4), blockSize);
        BinaryPrimitives.WriteUInt32LittleEndian(payload.AsSpan(12, 4), offset);
        BinaryPrimitives.WriteUInt32LittleEndian(payload.AsSpan(16, 4), encrypted ? 1u : 0u);

        return BuildPacket(ESP32Command.FlashBegin, payload);
    }

    /// <summary>
    /// Build FLASH_DATA command packet
    /// </summary>
    public static byte[] BuildFlashDataCommand(byte[] blockData, uint sequenceNumber)
    {
        var payload = new byte[16 + blockData.Length];

        BinaryPrimitives.WriteUInt32LittleEndian(payload.AsSpan(0, 4), (uint)blockData.Length);
        BinaryPrimitives.WriteUInt32LittleEndian(payload.AsSpan(4, 4), sequenceNumber);
        // Reserved (8 bytes of zeros) - already zero-initialized
        Array.Copy(blockData, 0, payload, 16, blockData.Length);

        var checksum = CalculateChecksum(blockData);
        return BuildPacket(ESP32Command.FlashData, payload, checksum);
    }

    /// <summary>
    /// Build FLASH_END command packet
    /// </summary>
    public static byte[] BuildFlashEndCommand(bool reboot = true)
    {
        var payload = new byte[4];
        // 0 = reboot, 1 = stay in bootloader
        BinaryPrimitives.WriteUInt32LittleEndian(payload, reboot ? 0u : 1u);
        return BuildPacket(ESP32Command.FlashEnd, payload);
    }

    /// <summary>
    /// Build CHANGE_BAUDRATE command packet
    /// </summary>
    public static byte[] BuildChangeBaudCommand(uint newBaud, uint oldBaud = 0)
    {
        var payload = new byte[8];
        BinaryPrimitives.WriteUInt32LittleEndian(payload.AsSpan(0, 4), newBaud);
        BinaryPrimitives.WriteUInt32LittleEndian(payload.AsSpan(4, 4), oldBaud);
        return BuildPacket(ESP32Command.ChangeBaudRate, payload);
    }

    /// <summary>
    /// Build READ_REG command packet
    /// </summary>
    public static byte[] BuildReadRegCommand(uint address)
    {
        var payload = new byte[4];
        BinaryPrimitives.WriteUInt32LittleEndian(payload, address);
        return BuildPacket(ESP32Command.ReadReg, payload);
    }

    /// <summary>
    /// Build WRITE_REG command packet
    /// </summary>
    public static byte[] BuildWriteRegCommand(
        uint address,
        uint value,
        uint mask = 0xFFFFFFFF,
        uint delayUs = 0)
    {
        var payload = new byte[16];
        BinaryPrimitives.WriteUInt32LittleEndian(payload.AsSpan(0, 4), address);
        BinaryPrimitives.WriteUInt32LittleEndian(payload.AsSpan(4, 4), value);
        BinaryPrimitives.WriteUInt32LittleEndian(payload.AsSpan(8, 4), mask);
        BinaryPrimitives.WriteUInt32LittleEndian(payload.AsSpan(12, 4), delayUs);
        return BuildPacket(ESP32Command.WriteReg, payload);
    }
}

/// <summary>
/// ESP32 bootloader response parser
/// </summary>
public class ESP32Response
{
    public byte Direction { get; init; }
    public byte Command { get; init; }
    public ushort Size { get; init; }
    public uint Value { get; init; }
    public byte[] Data { get; init; } = Array.Empty<byte>();
    public byte Status { get; init; }
    public byte Error { get; init; }

    public bool IsSuccess => Status == 0 && Error == 0;

    /// <summary>
    /// Parse a decoded SLIP packet into a response
    /// </summary>
    public static ESP32Response? Parse(byte[] packet)
    {
        if (packet.Length < 8) return null;

        var direction = packet[0];
        // Response direction should be 0x01
        if (direction != 0x01) return null;

        var command = packet[1];
        var size = BinaryPrimitives.ReadUInt16LittleEndian(packet.AsSpan(2, 2));
        var value = BinaryPrimitives.ReadUInt32LittleEndian(packet.AsSpan(4, 4));

        // Extract data (if any) and status bytes
        var dataLength = size > 2 ? size - 2 : 0;
        var data = new byte[dataLength];
        if (dataLength > 0 && packet.Length >= 8 + dataLength)
        {
            Array.Copy(packet, 8, data, 0, dataLength);
        }

        var statusIndex = 8 + dataLength;
        var status = packet.Length > statusIndex ? packet[statusIndex] : (byte)0;
        var error = packet.Length > statusIndex + 1 ? packet[statusIndex + 1] : (byte)0;

        return new ESP32Response
        {
            Direction = direction,
            Command = command,
            Size = size,
            Value = value,
            Data = data,
            Status = status,
            Error = error
        };
    }
}
